content/oci/deletable.go

```diff
@@ -0,0 +1,163 @@
+/*
+Copyright The ORAS Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package oci provides access to an OCI content store.
+// Reference: https://github.com/opencontainers/image-spec/blob/v1.1.0-rc4/image-layout.md
+package oci
+
+import (
+	"context"
+	"io"
+	"sync"
+
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"oras.land/oras-go/v2/content"
+)
+
+// DeletableStore implements `oras.Target`, and represents a content store
+// extended with the delete operation.
+// Reference: https://github.com/opencontainers/image-spec/blob/v1.1.0-rc4/image-layout.md
+type DeletableStore struct {
+	*Store
+	lock sync.RWMutex
+}
```

Comment:
~~Just a note: I find that any `Store` can be converted into a `DeletableStore`.~~
```go
store, err := NewStore("test")
handle(err)
deletableStore := &DeletableStore{
   Store: store,
}
```
~~It should just work :D like the constructor.~~

Reply:
I find it is dangerous for users to directly access the underlying `*Store`.

You may consider
```suggestion
type DeletableStore struct {
	AutoSaveIndex bool
	store *Store
	lock sync.RWMutex
}
```
and sync `AutoSaveIndex` to `store` in any write operations.

Reply:
We have decided to discard the implementation based on inheritance, and instead just extend `Store` with `Delete()` using a `RWMutex`.